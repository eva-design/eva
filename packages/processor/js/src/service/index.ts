export * from './common';
export * from './mapping';
export * from './style';
