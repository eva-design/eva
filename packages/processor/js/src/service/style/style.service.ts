import {
  ThemeMappingType,
  ControlMappingType,
  ThemedStyleType,
  MappingType,
  StrictTheme,
  StatelessMappingType,
  ParameterType,
} from '@eva-design/dss';
import {
  safe,
  noNulls,
  noDuplicates,
} from '../common';
import {
  getComponentDefaultAppearance,
  getComponentDefaultVariants,
  getComponentDefaultState,
  getStatelessAppearanceMapping,
  getStatelessVariantMapping,
  getStateAppearanceMapping,
  getStateVariantMapping,
} from '../mapping';

export const SEPARATOR_MAPPING_ENTRY = '.';

/**
 * Creates style object for variant/list of variants(optional) and its state/list of states(optional)
 *
 * Example
 *
 * appearance = 'outline';
 * variants = ['success', 'large'];
 * state = ['active', 'checked'];
 *
 * a = `default` + `outline`                    - acc appearance (apce) mapping
 *
 * v1 = `success` of `default`                  - `success` variant mapping of `default` apce
 * v2 = `success` of `outline`                  - `success` variant mapping of `outline` apce
 * v3 = `large` of `default`                    - `large` variant mapping of `default` apce
 * v4 = `large` of `outline`                    - `large` variant mapping of `outline` apce
 *
 * s1 = `active` of `default`                   - `active` state mapping of `default` apce
 * s2 = `active` of `outline`                   - `active` state mapping of `outline` apce
 * s3 = `active` of `default success`           - `active` state mapping of `success` variant of `default` apce
 * s4 = `active` of `outline success`           - `active` state mapping of `success` variant of `outline` apce
 * s5 = `active` of `default large`             - `active` state mapping of `large` variant of `default` apce
 * s6 = `active` of `outline large`             - `active` state mapping of `large` variant of `outline` apce
 *
 * s7 = `checked` of `default`                  - `checked` state mapping of `default` apce
 * s8 = `checked` of `outline`                  - `checked` state mapping of `outline` apce
 * s9 = `checked` of `default success`          - `checked` state mapping of `success` variant of `default` apce
 * s10 = `checked` of `outline success`         - `checked` state mapping of `success` variant of `outline` apce
 * s11 = `checked` of `default large`           - `checked` state mapping of `large` variant of `default` apce
 * s12 = `checked` of `outline large`           - `checked` state mapping of `large` variant of `outline` apce
 *
 * s13 = `active.checked` of `default`          - `active.checked` state mapping of `default` apce
 * s14 = `active.checked` of `outline`          - `active.checked` state mapping of `outline` apce
 * s15 = `active.checked` of `default success`  - `active.checked` state mapping of `success` variant of `default` apce
 * s16 = `active.checked` of `outline success`  - `active.checked` state mapping of `success` variant of `outline` apce
 * s17 = `active.checked` of `default large`    - `active.checked` state mapping of `large` variant of `default` apce
 * s18 = `active.checked` of `outline large`    - `active.checked` state mapping of `large` variant of `outline` apce
 *
 * res = a + (v1 + v2 + ... + vn) + (s1 + s2 + ... + sn)
 *
 * @param mapping: ThemeMappingType - theme mapping configuration
 * @param component: string - component name
 * @param appearance: string - appearance applied to component
 * @param variants: string[] - variants applied to component. Default is []
 * @param states: string[] - states in which component is. Default is []
 * @param theme: StrictTheme - theme configuration object. Default is {}
 *
 * @return ThemedStyleType - compiled component styles declared in mappings, mapped to theme values
 */
export function createStyle(mapping: ThemeMappingType,
                            component: string,
                            appearance: string,
                            variants: string[] = [],
                            states: string[] = [],
                            theme: StrictTheme = {}): ThemedStyleType {

  const normalizedAppearance: string[] = normalizeAppearance(mapping, component, appearance);
  const normalizedVariants: string[] = normalizeVariants(mapping, component, variants);
  const normalizedStates: string[] = normalizeStates(mapping, component, states, (state: string): number => {
    return states.indexOf(state);
  });

  const appearanceMapping: StatelessMappingType = reduce(normalizedAppearance, apce => {
    return getStatelessAppearanceMapping(mapping, component, apce);
  });

  const variantMapping: StatelessMappingType = reduce(normalizedVariants, variant => {
    return reduce(normalizedAppearance, apce => {
      return getStatelessVariantMapping(mapping, component, apce, variant);
    });
  });

  const stateMapping: StatelessMappingType = reduce(normalizedStates, state => {
    const appearanceStateMapping = reduce(normalizedAppearance, apce => {
      return getStateAppearanceMapping(mapping, component, apce, state);
    });

    const variantStateMapping: StatelessMappingType = reduce(normalizedVariants, variant => {
      return reduce(normalizedAppearance, apce => {
        return getStateVariantMapping(mapping, component, apce, variant, state);
      });
    });

    return { ...appearanceStateMapping, ...variantStateMapping };
  });

  const strictlessMapping: StatelessMappingType = {
    ...appearanceMapping,
    ...variantMapping,
    ...stateMapping,
  };

  return withStrictTokens(strictlessMapping, theme);
}

export function createAllStyles(mapping: ThemeMappingType,
                                component: string,
                                appearance: string,
                                variants: string[],
                                states: string[],
                                theme: StrictTheme): [string, ThemedStyleType][] {

  const stateless = createStyleEntry(mapping,
    component,
    appearance,
    appearance,
    '',
    '',
    theme,
  );

  const withStates = states.reduce((acc: [string, ThemedStyleType][], current: string) => {
    const key = appearance.concat(SEPARATOR_MAPPING_ENTRY, current);
    const next = createStyleEntry(mapping, component, key, appearance, '', current, theme);
    return [...acc, next];
  }, []);

  const withVariants = variants.map(variant => {
    const key = appearance.concat(SEPARATOR_MAPPING_ENTRY, variant);
    return createStyleEntry(mapping, component, key, appearance, variant, '', theme);
  });

  const withVariantStates = variants.reduce((acc: [string, ThemedStyleType][], current: string) => {
    const next = states.map(state => {
      const key = appearance.concat(SEPARATOR_MAPPING_ENTRY, current, SEPARATOR_MAPPING_ENTRY, state);
      return createStyleEntry(mapping, component, key, appearance, current, state, theme);
    });
    return [...acc, ...next];
  }, []);

  return [
    stateless,
    ...withStates,
    ...withVariants,
    ...withVariantStates,
  ];
}

export function getStyle(mapping: ThemeMappingType,
                         component: string,
                         appearance: string,
                         variants: string[],
                         states: string[]): ThemedStyleType | undefined {

  return safe(mapping, (themeMapping: ThemeMappingType) => {
    return safe(themeMapping[component], (componentMapping: ControlMappingType) => {

      const query = findStyleKey(Object.keys(componentMapping), [
        appearance,
        ...variants,
        ...states,
      ]);

      return componentMapping[query];
    });
  });
}

/**
 * Creates normalized to design system array of component appearances
 *
 * Example:
 *
 * '' => ['default']
 * 'bold' => ['default', 'bold']
 * 'default' => ['default']
 * ...
 *
 * @param mapping: ThemeMappingType - theme mapping configuration
 * @param component: string - component name
 * @param appearance: string - appearance applied to component
 *
 * @return string[] - array of merged appearances
 */
export function normalizeAppearance(mapping: ThemeMappingType,
                                    component: string,
                                    appearance: string): string[] {

  const defaultAppearance: string = getComponentDefaultAppearance(mapping, component);

  return normalize([defaultAppearance, appearance]);
}

/**
 * Creates normalized to design system array of component variants
 *
 * Example:
 *
 * [''] => ['default0', 'default1']
 * ['success'] => ['default0', 'default1', 'success']
 * ['default0', 'tiny'] => ['default0', 'default1', 'tiny']
 * ...
 *
 * @param mapping: ThemeMappingType - theme mapping configuration
 * @param component: string - component name
 * @param variants: string[] - variants applied to component
 *
 * @return string[] - array of merged variants
 */
export function normalizeVariants(mapping: ThemeMappingType,
                                  component: string,
                                  variants: string[]): string[] {

  const defaultVariants: string[] = getComponentDefaultVariants(mapping, component);

  return normalize([...defaultVariants, ...variants]);
}

/**
 * Creates normalized to design system array of component states
 *
 * Example:
 *
 * [''] => ['default']
 * ['active', 'checked'] => [
 *                           'default',
 *                           'active',
 *                           'default.active',
 *                           'checked',
 *                           'default.active.checked'
 *                          ]
 * ...
 *
 * @param mapping: ThemeMappingType - theme mapping configuration
 * @param component: string - component name
 * @param states: string[] - states in which component is
 * @param stateWeight: (state: string) => number - state weight calculation lambda
 * @param separator - state separator. `.` in example
 *
 * @return string[] - array of merged states
 */
export function normalizeStates(mapping: ThemeMappingType,
                                component: string,
                                states: string[],
                                stateWeight: (state: string) => number,
                                separator: string = SEPARATOR_MAPPING_ENTRY): string[] {

  const defaultState: string = getComponentDefaultState(mapping, component);

  const preprocess: string[] = normalize([defaultState, ...states]);
  if (preprocess.length === 0) {
    return preprocess;
  }

  const variations = createStateVariations([...preprocess], separator, []);

  return variations.sort((lhs: string, rhs: string) => {
    const lhsWeight = getStateVariationWeight(lhs, separator, stateWeight);
    const rhsWeight = getStateVariationWeight(rhs, separator, stateWeight);
    return lhsWeight - rhsWeight;
  });
}

function createStateVariations(states: string[], separator: string, result: string[] = []): string[] {
  if (states.length === 0) {
    return result;
  }

  const next = states.reduce((acc: string[], current: string) => {
    const nextItem = acc.map(value => value.concat(separator, current));
    return [...acc, ...nextItem];
  }, [states.shift()]);

  return createStateVariations(states, separator, [...result, ...next]);
}

function withStrictTokens(mapping: StatelessMappingType, theme: StrictTheme): StatelessMappingType {
  return Object.keys(mapping).reduce((acc: StatelessMappingType, next: string): StatelessMappingType => {
    const currentToken: ParameterType = mapping[next];
    const nextToken: ParameterType = theme[currentToken] || currentToken;

    return { ...acc, [next]: nextToken };
  }, {});
}

function getStateVariationWeight(state: string,
                                 separator: string,
                                 stateWeight: (state: string) => number): number {

  const parts = state.split(separator);
  return parts.reduce((acc: number, current: string): number => {
    return acc + stateWeight(current) + parts.length;
  }, 0);
}

/**
 * Finds identical keys across `source` keys array
 *
 * Example:
 *
 * source = ['default.error.small.checked', ...]
 * query = ['default', 'small', 'error', 'checked']
 *
 * will return 'default.error.small.checked'
 *
 * @param source (string[]) - array of style keys
 * @param query (string[]) - array of key parts to search
 *
 * @return (string | undefined) - key identical to some of `source` keys if presents
 */
export function findStyleKey(source: string[], query: string[]): string | undefined {

  const partialKeys: string[][] = source.map((key: string) => {
    return key.split(SEPARATOR_MAPPING_ENTRY);
  });

  const result: string[][] = partialKeys.filter((partial: string[]) => {
    return compareArrays(query, partial);
  });

  return safe(result[0], (value: string[]) => {
    return value.join(SEPARATOR_MAPPING_ENTRY);
  });
}

function createStyleEntry(mapping: ThemeMappingType,
                          component: string,
                          key: string,
                          appearance: string,
                          variant: string = '',
                          state: string = '',
                          theme: StrictTheme = {}): [string, ThemedStyleType] {

  const value = createStyle(
    mapping,
    component,
    appearance,
    variant.split(SEPARATOR_MAPPING_ENTRY),
    state.split(SEPARATOR_MAPPING_ENTRY),
    theme,
  );

  return [key, value];
}

function normalize(params: string[]): string[] {
  return noNulls(noDuplicates(params));
}

function compareArrays(lhs: string[], rhs: string[]): boolean {
  const isEqualLength = lhs && rhs && lhs.length === rhs.length;
  if (!isEqualLength) {
    return false;
  }

  return lhs.reduce((acc: boolean, next: string): boolean => acc && rhs.includes(next), true);
}

function reduce(items: string[], next: (item: string) => any): any {
  return items.reduce((acc, current) => ({ ...acc, ...next(current) }), {});
}
