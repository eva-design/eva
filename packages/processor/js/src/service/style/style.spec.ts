import * as Service from './style.service';
import { mapping } from './style.spec.config';
import {
  ThemedStyleType,
  StrictTheme,
} from '@eva-design/dss';

describe('@style: service methods checks', () => {

  describe('* preprocess', () => {

    describe('* normalizes appearance properly', () => {

      it('* explicit default', () => {
        const value: string[] = Service.normalizeAppearance(
          mapping,
          'Button',
          'filled',
        );

        expect(value).toEqual([
          'filled',
        ]);
      });

      it('* custom', () => {
        const value: string[] = Service.normalizeAppearance(
          mapping,
          'Button',
          'outline',
        );

        expect(value).toEqual([
          'filled',
          'outline',
        ]);
      });

      it('* empty', () => {
        const value: string[] = Service.normalizeAppearance(
          mapping,
          'Button',
          '',
        );

        expect(value).toEqual([
          'filled',
        ]);
      });

      it('* undefined', () => {
        const value: string[] = Service.normalizeAppearance(
          mapping,
          'Button',
          undefined,
        );

        expect(value).toEqual([
          'filled',
        ]);
      });

    });

    describe('* normalizes variants properly', () => {

      it('* explicit default', () => {
        const value: string[] = Service.normalizeVariants(
          mapping,
          'Button',
          ['primary', 'medium'],
        );

        expect(value).toEqual([
          'primary',
          'medium',
        ]);
      });

      it('* with custom', () => {
        const value: string[] = Service.normalizeVariants(
          mapping,
          'Button',
          ['success'],
        );

        expect(value).toEqual([
          'primary',
          'medium',
          'success',
        ]);
      });

      it('* empty', () => {
        const value: string[] = Service.normalizeVariants(
          mapping,
          'Button',
          [],
        );

        expect(value).toEqual([
          'primary',
          'medium',
        ]);
      });

      it('* with duplicates', () => {
        const value: string[] = Service.normalizeVariants(
          mapping,
          'Button',
          ['success', 'success'],
        );

        expect(value).toEqual([
          'primary',
          'medium',
          'success',
        ]);
      });

      it('* with undefined', () => {
        const value: string[] = Service.normalizeVariants(
          mapping,
          'Button',
          [undefined, null],
        );

        expect(value).toEqual([
          'primary',
          'medium',
        ]);
      });

    });

    describe('* normalizes states properly', () => {
      const states: string[] = [
        'active',
        'checked',
        'disabled',
      ];

      const calculateStateWeight = (state: string): number => {
        switch (state) {
          case 'active':
            return 0;
          case 'checked':
            return 1;
          case 'disabled':
            return 2;
        }
      };

      it('* with applied states', () => {
        const value: string[] = Service.normalizeStates(
          mapping,
          'Button',
          states,
          calculateStateWeight,
        );

        expect(value).toEqual([
          'active',
          'checked',
          'disabled',
          'active.checked',
          'active.disabled',
          'checked.disabled',
          'active.checked.disabled',
        ]);
      });

      it('* empty', () => {
        const value: string[] = Service.normalizeStates(
          mapping,
          'Button',
          [],
          calculateStateWeight,
        );

        expect(value).toEqual([]);
      });

      it('* with duplicates', () => {
        const value: string[] = Service.normalizeStates(
          mapping,
          'Button',
          [...states, 'active'],
          calculateStateWeight,
        );

        expect(value).toEqual([
          'active',
          'checked',
          'disabled',
          'active.checked',
          'active.disabled',
          'checked.disabled',
          'active.checked.disabled',
        ]);
      });

      it('* with undefined', () => {
        const value: string[] = Service.normalizeStates(
          mapping,
          'Button',
          [...states, undefined],
          calculateStateWeight,
        );

        expect(value).toEqual([
          'active',
          'checked',
          'disabled',
          'active.checked',
          'active.disabled',
          'checked.disabled',
          'active.checked.disabled',
        ]);
      });

      it('* custom separator', () => {
        const value: string[] = Service.normalizeStates(
          mapping,
          'Button',
          states,
          calculateStateWeight,
          '-',
        );

        expect(value).toEqual([
          'active',
          'checked',
          'disabled',
          'active-checked',
          'active-disabled',
          'checked-disabled',
          'active-checked-disabled',
        ]);
      });

    });

    describe('* style query', () => {

      const source: string[] = [
        'default',
        'default.checked',
        'default.success',
        'default.success.checked',
        'default.success.small.checked',
        'default.success.small.checked.active',
      ];

      it('* appearance only', () => {
        const value: string = Service.findStyleKey(source, [
          'default',
        ]);

        expect(value).toEqual('default');
      });

      it('* appearance and state', () => {
        const value: string = Service.findStyleKey(source, [
          'default',
          'checked',
        ]);

        expect(value).toEqual('default.checked');
      });

      it('* appearance and variant', () => {
        const value: string = Service.findStyleKey(source, [
          'default',
          'success',
        ]);

        expect(value).toEqual('default.success');
      });

      it('* appearance and variant and state', () => {
        const value: string = Service.findStyleKey(source, [
          'default',
          'success',
          'checked',
        ]);

        expect(value).toEqual('default.success.checked');
      });

      it('* appearance and variants and state', () => {
        const value: string = Service.findStyleKey(source, [
          'default',
          'success',
          'small',
          'checked',
        ]);

        expect(value).toEqual('default.success.small.checked');
      });

      it('* appearance and variants and states', () => {
        const value: string = Service.findStyleKey(source, [
          'default',
          'success',
          'small',
          'checked',
          'active',
        ]);

        expect(value).toEqual('default.success.small.checked.active');
      });

      it('* unordered', () => {
        const value: string = Service.findStyleKey(source, [
          'default',
          'checked',
          'small',
          'success',
          'active',
        ]);

        expect(value).toEqual('default.success.small.checked.active');
      });

      it('* with undefined in config', () => {
        const value: string = Service.findStyleKey(source, [
          'default',
          'error',
          'small',
          'checked',
          'active',
        ]);

        expect(value).toEqual(undefined);
      });

    });

  });

  describe('* styling', () => {

    describe('* default appearance', () => {

      const appearance: string = 'filled';

      it('* stateless', () => {
        const value: ThemedStyleType = Service.createStyle(mapping, 'Button', appearance);

        expect(value).toMatchSnapshot();
      });

      describe('* with state', () => {

        it('* single', () => {
          const value: ThemedStyleType = Service.createStyle(
            mapping,
            'Button',
            appearance,
            [],
            ['active'],
          );

          expect(value).toMatchSnapshot();
        });

        it('* multiple', () => {
          const value: ThemedStyleType = Service.createStyle(
            mapping,
            'Button',
            appearance,
            [],
            ['disabled', 'active'],
          );

          expect(value).toMatchSnapshot();
        });
      });

      describe('* with variant', () => {

        describe('* single', () => {

          it('* no state', () => {
            const value: ThemedStyleType = Service.createStyle(
              mapping,
              'Button',
              appearance,
              ['success'],
            );

            expect(value).toMatchSnapshot();
          });

          describe('* with state', () => {

            it('* single implicit (should apply from appearance)', () => {
              const value: ThemedStyleType = Service.createStyle(
                mapping,
                'Button',
                appearance,
                ['success'],
                ['active'],
              );

              expect(value).toMatchSnapshot();
            });

            it('* single explicit (should apply own)', () => {
              const value: ThemedStyleType = Service.createStyle(
                mapping,
                'Button',
                appearance,
                ['success'],
                ['active'],
              );

              expect(value).toMatchSnapshot();
            });

            it('* multiple', () => {
              const value: ThemedStyleType = Service.createStyle(
                mapping,
                'Button',
                appearance,
                ['success'],
                ['disabled', 'active'],
              );

              expect(value).toMatchSnapshot();
            });

          });

        });

        describe('* multiple', () => {

          it('* no state', () => {
            const value: ThemedStyleType = Service.createStyle(
              mapping,
              'Button',
              appearance,
              ['success', 'large'],
            );

            expect(value).toMatchSnapshot();
          });

          describe('* with state', () => {

            it('* single', () => {
              const value: ThemedStyleType = Service.createStyle(
                mapping,
                'Button',
                appearance,
                ['success', 'large'],
                ['active'],
              );

              expect(value).toMatchSnapshot();
            });

            it('* multiple', () => {
              const value: ThemedStyleType = Service.createStyle(
                mapping,
                'Button',
                appearance,
                ['success', 'large'],
                ['disabled', 'active'],
              );

              expect(value).toMatchSnapshot();
            });

          });

        });

      });

      describe('* strict tokens', () => {

        const textPrimaryInverseValue: string = 'white';

        const strict: StrictTheme = {
          'text-primary-inverse': textPrimaryInverseValue,
        };

        it('* maps strict tokens to corresponding values', () => {
          const value: ThemedStyleType = Service.createStyle(
            mapping,
            'Button',
            appearance,
            [],
            [],
            strict,
          );

          const { textColor } = value;

          expect(textColor).toEqual(textPrimaryInverseValue);
        });
      });

    });

    describe('* custom appearance', () => {

      const appearance: string = 'outline';

      it('* stateless', () => {
        const value: ThemedStyleType = Service.createStyle(
          mapping,
          'Button',
          'outline',
        );

        expect(value).toMatchSnapshot();
      });

      describe('* with state', () => {

        it('* implicit (should apply from default appearance)', () => {
          const value: ThemedStyleType = Service.createStyle(
            mapping,
            'Button',
            appearance,
            [],
            ['disabled'],
          );

          expect(value).toMatchSnapshot();
        });

        it('* explicit (should apply own)', () => {
          const value: ThemedStyleType = Service.createStyle(
            mapping,
            'Button',
            appearance,
            [],
            ['active'],
          );

          expect(value).toMatchSnapshot();
        });

      });

      describe('* with variant', () => {

        it('* implicit (should apply from default appearance)', () => {
          const value: ThemedStyleType = Service.createStyle(
            mapping,
            'Button',
            appearance,
            ['large'],
          );

          expect(value).toMatchSnapshot();
        });

        it('* explicit (should apply own)', () => {
          const value: ThemedStyleType = Service.createStyle(
            mapping,
            'Button',
            appearance,
            ['success'],
          );

          expect(value).toMatchSnapshot();
        });

      });

    });

    describe('* undefined appearance', () => {

      const appearance: string = 'undefined';

      it('* stateless (should apply default appearance)', () => {
        const value: ThemedStyleType = Service.createStyle(
          mapping,
          'Button',
          appearance,
        );

        expect(value).toMatchSnapshot();
      });

    });

  });

});
