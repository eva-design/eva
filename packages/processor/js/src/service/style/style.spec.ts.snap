// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`@style: service methods checks * styling * custom appearance * stateless 1`] = `
Object {
  "backgroundColor": "transparent",
  "borderColor": "#2196F3",
  "borderRadius": 6,
  "borderWidth": 2,
  "iconHeight": 13,
  "iconMarginHorizontal": 4,
  "iconTintColor": "#2196F3",
  "iconWidth": 13,
  "minHeight": 24,
  "minWidth": 24,
  "padding": 4,
  "textColor": "#2196F3",
  "textFontSize": 12,
  "textFontWeight": 800,
  "textMarginHorizontal": 4,
}
`;

exports[`@style: service methods checks * styling * custom appearance * with state * explicit (should apply own) 1`] = `
Object {
  "backgroundColor": "transparent",
  "borderColor": "#1E88E5",
  "borderRadius": 6,
  "borderWidth": 2,
  "iconHeight": 13,
  "iconMarginHorizontal": 4,
  "iconTintColor": "#1E88E5",
  "iconWidth": 13,
  "minHeight": 24,
  "minWidth": 24,
  "padding": 4,
  "textColor": "#1E88E5",
  "textFontSize": 12,
  "textFontWeight": 800,
  "textMarginHorizontal": 4,
}
`;

exports[`@style: service methods checks * styling * custom appearance * with state * implicit (should apply from default appearance) 1`] = `
Object {
  "backgroundColor": "transparent",
  "borderColor": "#2196F3",
  "borderRadius": 6,
  "borderWidth": 2,
  "iconHeight": 13,
  "iconMarginHorizontal": 4,
  "iconTintColor": "#2196F3",
  "iconWidth": 13,
  "minHeight": 24,
  "minWidth": 24,
  "padding": 4,
  "textColor": "#2196F3",
  "textFontSize": 12,
  "textFontWeight": 800,
  "textMarginHorizontal": 4,
}
`;

exports[`@style: service methods checks * styling * custom appearance * with variant * explicit (should apply own) 1`] = `
Object {
  "backgroundColor": "transparent",
  "borderColor": "#4CAF50",
  "borderRadius": 6,
  "borderWidth": 2,
  "iconHeight": 13,
  "iconMarginHorizontal": 4,
  "iconTintColor": "#4CAF50",
  "iconWidth": 13,
  "minHeight": 24,
  "minWidth": 24,
  "padding": 4,
  "textColor": "#4CAF50",
  "textFontSize": 12,
  "textFontWeight": 800,
  "textMarginHorizontal": 4,
}
`;

exports[`@style: service methods checks * styling * custom appearance * with variant * implicit (should apply from default appearance) 1`] = `
Object {
  "backgroundColor": "transparent",
  "borderColor": "#2196F3",
  "borderRadius": 6,
  "borderWidth": 2,
  "iconHeight": 14,
  "iconMarginHorizontal": 4.5,
  "iconTintColor": "#2196F3",
  "iconWidth": 14,
  "minHeight": 30,
  "minWidth": 30,
  "padding": 4.5,
  "textColor": "#2196F3",
  "textFontSize": 13,
  "textFontWeight": 800,
  "textMarginHorizontal": 4.5,
}
`;

exports[`@style: service methods checks * styling * default appearance * stateless 1`] = `
Object {
  "backgroundColor": "#2196F3",
  "borderRadius": 6,
  "iconHeight": 13,
  "iconMarginHorizontal": 4,
  "iconTintColor": "#ffffff",
  "iconWidth": 13,
  "minHeight": 24,
  "minWidth": 24,
  "padding": 4,
  "textColor": "text-primary-inverse",
  "textFontSize": 12,
  "textFontWeight": 800,
  "textMarginHorizontal": 4,
}
`;

exports[`@style: service methods checks * styling * default appearance * with state * multiple 1`] = `
Object {
  "backgroundColor": "#1E88E5",
  "borderRadius": 6,
  "iconHeight": 13,
  "iconMarginHorizontal": 4,
  "iconTintColor": "#ffffff",
  "iconWidth": 13,
  "minHeight": 24,
  "minWidth": 24,
  "padding": 4,
  "textColor": "text-primary-inverse",
  "textFontSize": 12,
  "textFontWeight": 800,
  "textMarginHorizontal": 4,
}
`;

exports[`@style: service methods checks * styling * default appearance * with state * single 1`] = `
Object {
  "backgroundColor": "#1E88E5",
  "borderRadius": 6,
  "iconHeight": 13,
  "iconMarginHorizontal": 4,
  "iconTintColor": "#ffffff",
  "iconWidth": 13,
  "minHeight": 24,
  "minWidth": 24,
  "padding": 4,
  "textColor": "text-primary-inverse",
  "textFontSize": 12,
  "textFontWeight": 800,
  "textMarginHorizontal": 4,
}
`;

exports[`@style: service methods checks * styling * default appearance * with variant * multiple * no state 1`] = `
Object {
  "backgroundColor": "#4CAF50",
  "borderRadius": 6,
  "iconHeight": 14,
  "iconMarginHorizontal": 4.5,
  "iconTintColor": "#ffffff",
  "iconWidth": 14,
  "minHeight": 30,
  "minWidth": 30,
  "padding": 4.5,
  "textColor": "text-primary-inverse",
  "textFontSize": 13,
  "textFontWeight": 800,
  "textMarginHorizontal": 4.5,
}
`;

exports[`@style: service methods checks * styling * default appearance * with variant * multiple * with state * multiple 1`] = `
Object {
  "backgroundColor": "#43A047",
  "borderRadius": 6,
  "iconHeight": 14,
  "iconMarginHorizontal": 4.5,
  "iconTintColor": "#ffffff",
  "iconWidth": 14,
  "minHeight": 30,
  "minWidth": 30,
  "padding": 4.5,
  "textColor": "text-primary-inverse",
  "textFontSize": 13,
  "textFontWeight": 800,
  "textMarginHorizontal": 4.5,
}
`;

exports[`@style: service methods checks * styling * default appearance * with variant * multiple * with state * single 1`] = `
Object {
  "backgroundColor": "#43A047",
  "borderRadius": 6,
  "iconHeight": 14,
  "iconMarginHorizontal": 4.5,
  "iconTintColor": "#ffffff",
  "iconWidth": 14,
  "minHeight": 30,
  "minWidth": 30,
  "padding": 4.5,
  "textColor": "text-primary-inverse",
  "textFontSize": 13,
  "textFontWeight": 800,
  "textMarginHorizontal": 4.5,
}
`;

exports[`@style: service methods checks * styling * default appearance * with variant * single * no state 1`] = `
Object {
  "backgroundColor": "#4CAF50",
  "borderRadius": 6,
  "iconHeight": 13,
  "iconMarginHorizontal": 4,
  "iconTintColor": "#ffffff",
  "iconWidth": 13,
  "minHeight": 24,
  "minWidth": 24,
  "padding": 4,
  "textColor": "text-primary-inverse",
  "textFontSize": 12,
  "textFontWeight": 800,
  "textMarginHorizontal": 4,
}
`;

exports[`@style: service methods checks * styling * default appearance * with variant * single * with state * multiple 1`] = `
Object {
  "backgroundColor": "#43A047",
  "borderRadius": 6,
  "iconHeight": 13,
  "iconMarginHorizontal": 4,
  "iconTintColor": "#ffffff",
  "iconWidth": 13,
  "minHeight": 24,
  "minWidth": 24,
  "padding": 4,
  "textColor": "text-primary-inverse",
  "textFontSize": 12,
  "textFontWeight": 800,
  "textMarginHorizontal": 4,
}
`;

exports[`@style: service methods checks * styling * default appearance * with variant * single * with state * single explicit (should apply own) 1`] = `
Object {
  "backgroundColor": "#43A047",
  "borderRadius": 6,
  "iconHeight": 13,
  "iconMarginHorizontal": 4,
  "iconTintColor": "#ffffff",
  "iconWidth": 13,
  "minHeight": 24,
  "minWidth": 24,
  "padding": 4,
  "textColor": "text-primary-inverse",
  "textFontSize": 12,
  "textFontWeight": 800,
  "textMarginHorizontal": 4,
}
`;

exports[`@style: service methods checks * styling * default appearance * with variant * single * with state * single implicit (should apply from appearance) 1`] = `
Object {
  "backgroundColor": "#43A047",
  "borderRadius": 6,
  "iconHeight": 13,
  "iconMarginHorizontal": 4,
  "iconTintColor": "#ffffff",
  "iconWidth": 13,
  "minHeight": 24,
  "minWidth": 24,
  "padding": 4,
  "textColor": "text-primary-inverse",
  "textFontSize": 12,
  "textFontWeight": 800,
  "textMarginHorizontal": 4,
}
`;

exports[`@style: service methods checks * styling * undefined appearance * stateless (should apply default appearance) 1`] = `
Object {
  "backgroundColor": "#2196F3",
  "borderRadius": 6,
  "iconHeight": 13,
  "iconMarginHorizontal": 4,
  "iconTintColor": "#ffffff",
  "iconWidth": 13,
  "minHeight": 24,
  "minWidth": 24,
  "padding": 4,
  "textColor": "text-primary-inverse",
  "textFontSize": 12,
  "textFontWeight": 800,
  "textMarginHorizontal": 4,
}
`;
